aten/src/ATen/native/cuda/DepthwiseConv2d.cu

```diff
@@ -1,29 +1,597 @@
+    // the other parameters we have
+      int go_w_offset = idx % outputWidth;
+      int go_h_offset = (idx / outputWidth);
+
+      int i_w_offset = (go_w_offset * strideWidth) + (kW * dilationWidth) - padWidth;
+      int i_h_offset = (go_h_offset * strideHeight) + (kH * dilationHeight) - padHeight;
+
+      if (i_w_offset >= 0 && i_h_offset >= 0 && i_w_offset < inputWidth && i_h_offset < inputHeight) {
+        int inputOffset = ((batchIdx * inputChannels + inputCh) * inputHeight + i_h_offset) * inputWidth + i_w_offset;
+        int outputOffset = ((batchIdx * kernelChannels + ch) * outputHeight ) * outputWidth + idx;
+        grad += (static_cast<acc_t>(input.data()[inputOffset]) *
+                 static_cast<acc_t>(grad_output.data()[outputOffset]));
+      }
+    }
+  }
+  __syncthreads();
+
+  // At this point each thread in the block has a local gradient, which we need to
+  // accumulate prior to writing the global value
+  extern __shared__ char smem[];
+  acc_t* buf = reinterpret_cast<acc_t*>(smem);
+  acc_t tval = reduceBlock(buf, blockDim.x, grad,
+                           [] __device__ (acc_t a, acc_t b) { return a + b; },
+                           acc_t(0));
+
+  // After reduction, first thread in the block has the gradient, so its responsible
+  // for writing it to grad_weight
+  if (threadIdx.x == 0) {
+    int weightOffset = kW + (kernelWidth * kH) + (kernelWidth * kernelHeight * ch);
+    grad_weight.data()[weightOffset] = static_cast<scalar_t>(tval);
+  }
+}
+
+void conv_depthwise2d_forward_out(
+                  const Tensor &input,
+                  const Tensor &output,
+                  const Tensor &weight,
+                  const Tensor &bias,
+                  int kW, int kH,
+                  int dW, int dH,
+                  int padW, int padH,
+                  int dilationW, int dilationH) {
+  // Only handle 4D Input Tensors for now
+  TORCH_CHECK(input.numel() > 0 && input.dim() == 4);
+  TORCH_CHECK(weight.numel() > 0 && weight.dim() == 4);
+  TORCH_CHECK(output.is_contiguous());
+
+  auto in_sizes = input.sizes();
+  auto w_sizes = weight.sizes();
+
+  // We assume that the input and weight Tensors are shaped properly by
+  // the caller, so we verify that here to some extent
+
+  // Weight Tensor is shape (output_channels, 1, kH, kW)
+  TORCH_CHECK(w_sizes[1] == 1);
+
+  // Input Tensor is shape (N, input_channels, H, W)
+  // We verify that the # of output_channels is a multiple of input_channels
+  TORCH_CHECK(w_sizes[0] % in_sizes[1] == 0);
+
+  // Bias has same # of channels as output
+  const bool has_bias = bias.defined();
+  TORCH_CHECK(!has_bias || (bias.dim() <= 1 && bias.numel() == w_sizes[0]));
+
+  // Following the behavior of other THCUNN functions, we shape the output
+  // Tensor ourselves
+  int64_t batchSize = in_sizes[0];
+  int64_t height = in_sizes[2];
+  int64_t width = in_sizes[3];
+  int64_t outputHeight = (height + 2 * padH - (dilationH * (kH - 1) + 1)) / dH + 1;
```

Comment:
nit: you can use `conv_output_size` here (to make sure it doesn't diverge with how other convolution implementations define it)